src/utils/canister/actor.ts

```diff
@@ -15,20 +15,62 @@ function getHost() {
 
 const host = getHost();
 
-// Since we're using webpack-dev-server as part of create-react-app, we need to
-// add its port to our HttpAgent config as the host.
-const agent = new HttpAgent({ host });
-export const baseActor = Actor.createActor<_SERVICE>(CanCan_idl, {
-  agent,
-  canisterId: CanCan_canister_id,
-});
-
-export class CanCanActor {
-  actor: _SERVICE;
-  constructor(overrideActor?: _SERVICE) {
-    this.actor = overrideActor ?? baseActor;
+function createActor(identity?: Identity) {
+  const agent = new HttpAgent({ host, identity });
+  const actor = Actor.createActor<_SERVICE>(CanCan_idl, {
+    agent,
+    canisterId: CanCan_canister_id,
+  });
+  return { actor, agent };
+}
+
+/*
+ * Responsible for keeping track of the actor, whether the user has logged
+ * in again or not. A logged in user uses a different actor with their
+ * Identity, to ensure their Principal is passed to the backend.
+ */
+class ActorController {
+  _actor: _SERVICE;
+  _isAuthenticated: boolean = false;
+  isReady: boolean = false;
+
+  constructor() {
+    this._actor = this.initBaseActor();
+  }
+
+  initBaseActor() {
+    const { agent, actor } = createActor();
+    agent.fetchRootKey().then(() => (this.isReady = true));
+    return actor;
+  }
+
+  /*
+   * Get the actor instance to run commands on the canister.
+   */
+  get actor() {
+    return this._actor;
+  }
+
+  /*
+   * Once a user has authenticated and has an identity pass this identity
+   * to create a new actor with it, so they pass their Principal to the backend.
+   */
+  async authenticateActor(identity: Identity) {
```

Comment:
Instead of sometimes having an `Identity` and sometimes not, this is why @hansl made `AnonymousIdentity` you can import from `@dfinity/agent`. Then there is always *some* identity object, it's just a matter of which kind (`instanceof`) or what distinct value of `identity.getPrincipal()`

Reply:
Cool idea, I can look into doing that